emg/emg_local_STFT_execute.py

```diff
@@ -0,0 +1,172 @@
+def calc_stft_mode_freq(dat, BSA_output = True, **stft_params):
+    """
+    Calculate the mode frequency of the STFT of a signal
+
+    Inputs:
+        dat : 1D array, signal to be analyzed
+        stft_params : dict, parameters for STFT calculation
+        BSA_output : bool, whether to output in a similar format to the BSA 
+
+    Outputs:
+        freq_vec : 1D array, frequency vector for STFT
+        t_vec : 1D array, time vector for STFT
+        weight_mean : 1D array, time averaged mode frequency
+    """
+    freq_vec, t_vec, stft  = calc_stft(dat, **stft_params)
+    mag = np.abs(stft)
+    time_norm_mag = mag / mag.sum(axis=0)
+    weight_mean = (freq_vec[:,None] * time_norm_mag).sum(axis=0)
+    if BSA_output:
+        # This is only needed if stft_step size > 1
+        # Interpolate output to same length as input
+        new_t_vec = np.arange(stft_params['time_range_tuple'][0],
+                              stft_params['time_range_tuple'][1],
+                              1/stft_params['Fs'])
+        new_weight_mean = np.interp(new_t_vec, t_vec, weight_mean)
+        new_freq = np.arange(0,10,0.5)
+        new_weight_mean = np.zeros((len(new_freq), len(t_vec)))
+        closest_freq = np.argmin(np.abs(weight_mean[:,None] - new_freq), axis=1)
+        new_weight_mean[closest_freq, np.arange(len(t_vec))] = 1
+        return new_freq, new_t_vec, new_weight_mean
+    else:
+        return freq_vec, t_vec, weight_mean
+
+
+##############################
+# Setup params
+##############################
+script_dir = os.path.dirname(os.path.realpath(__file__))
+blech_clust_dir = os.path.dirname(script_dir)
+emg_params_path = os.path.join(blech_clust_dir, 'params', 'emg_params.json')
+
+if not os.path.exists(emg_params_path):
+    print('=== Environment params file not found. ===')
+    print('==> Please copy [[ blech_clust/params/_templates/emg_params.json ]] to [[ blech_clust/params/env_params.json ]] and update as needed.')
+    exit()
+
+emg_params_dict = json.load(open(emg_params_path, 'r'))
+stft_params = emg_params_dict['stft_params']
+
+
+##############################
+# Calculate STFT 
+##############################
+# Read blech.dir, and cd to that directory. 
+with open('BSA_run.dir', 'r') as f:
+    dir_list = [x.strip() for x in f.readlines()]
+
+# If there is more than one dir in BSA_run.dir,
+# loop over both, as both sets will have the same number of trials
+for dir_name in dir_list: 
+    sys.stdout = Logger(os.path.join(dir_name, 'BSA_log.txt'))
+    os.chdir(dir_name)
+
+    # Read the data files
+    emg_env = np.load('emg_env.npy')
+    sig_trials = np.load('sig_trials.npy')
+
+    # cd to emg_BSA_results
+    os.chdir('emg_BSA_results')
+
+    task = int(sys.argv[1])
+
+    taste = int((task-1)/sig_trials.shape[-1])
+    trial = int((task-1)%sig_trials.shape[-1])
+
+    print(f'Processing taste {taste}, trial {trial}')
+
+    # Make the time array and assign it to t on R
+    T = (np.arange(7000) + 1)/1000.0
```

Comment:
Hardcoding carries over from NM's code. Making changes to array size might cause issues downstream with BSA segmentation because segmentation uses indices rather than frequencies to pull out appropriate bands.

Will create new issue for this.